File: amp-python/src/amp/commands/update.py
from __future__ import annotations

import click

from ..types import STATUS_VALUES
from ._common import console, get_store, handle_amp_errors


@click.command()
@click.argument("node_id")
@click.option("--content", default=None, help="Replace the node's body content.")
@click.option("--tags", default=None, help="Replace tags (comma-separated).")
@click.option("--add-tags", default=None, help="Add tags (comma-separated).")
@click.option("--remove-tags", default=None, help="Remove tags (comma-separated).")
@click.option("--confidence", type=float, default=None, help="Update confidence score.")
@click.option("--status", type=click.Choice(STATUS_VALUES), default=None, help="Update status.")
@click.option("--scope", type=click.Choice(["agent", "user", "team", "workspace", "public"]),
              default=None, help="Update scope.")
@click.pass_context
@handle_amp_errors
def update_cmd(
    ctx: click.Context,
    node_id: str,
    content: str | None,
    tags: str | None,
    add_tags: str | None,
    remove_tags: str | None,
    confidence: float | None,
    status: str | None,
    scope: str | None,
) -> None:
    """Modify an existing node."""
    store = get_store(ctx.obj.get("path"))
    node = store.update(
        node_id,
        content=content,
        tags=[t.strip() for t in tags.split(",")] if tags else None,
        add_tags=[t.strip() for t in add_tags.split(",")] if add_tags else None,
        remove_tags=[t.strip() for t in remove_tags.split(",")] if remove_tags else None,
        confidence=confidence,
        status=status,
        scope=scope,
    )
    console.print(f"[green]Updated[/green] {node.id} (modified {node.modified})")

File: amp-cli/src/commands/update.ts
import { Command } from "commander";
import { requireNodeById, writeNode } from "../core/node.js";
import { appendChangelog, readManifest, requireStore } from "../core/store.js";
import { reindexStore } from "../core/index.js";
import { nowIso } from "../utils/helpers.js";
import { chalk, kv, success } from "../utils/ui.js";

interface UpdateOptions {
  content?: string;
  tags?: string;
  confidence?: string;
  status?: string;
  addTag?: string;
  removeTag?: string;
  noIndex?: boolean;
}

export function registerUpdateCommand(program: Command): void {
  program
    .command("update <id>")
    .description("Modify an existing memory node")
    .option("--content <content>", "Replace the node's body content")
    .option("--tags <tags>", "Replace all tags (comma-separated)")
    .option("--add-tag <tag>", "Add a single tag")
    .option("--remove-tag <tag>", "Remove a single tag")
    .option("--confidence <n>", "Update confidence score (0-1)")
    .option("--status <status>", "Update status")
    .option("--no-index", "Skip automatic reindexing")
    .action((id: string, options: UpdateOptions) => {
      const paths = requireStore(process.cwd());
      const manifest = readManifest(paths);
      const node = requireNodeById(paths, id);

      const changed: string[] = [];

      if (options.content !== undefined) {
        node.content = options.content.trim().startsWith("#")
          ? options.content.trim()
          : `# ${node.content.match(/^#\s+(.+)$/m)?.[1] ?? /* v8 ignore next */ id}\n\n${options.content.trim()}`;
        changed.push("content");
      }

      if (options.tags !== undefined) {
        node.frontmatter.tags = options.tags.split(",").map((t) => t.trim()).filter(Boolean);
        changed.push("tags");
      }
      if (options.addTag) {
        const tags = new Set(node.frontmatter.tags ?? []);
        tags.add(options.addTag.trim());
        node.frontmatter.tags = Array.from(tags);
        changed.push("tags");
      }
      if (options.removeTag) {
        node.frontmatter.tags = (node.frontmatter.tags ?? []).filter((t) => t !== options.removeTag);
        changed.push("tags");
      }

      if (options.confidence !== undefined) {
        const confidence = Number(options.confidence);
        if (Number.isNaN(confidence) || confidence < 0 || confidence > 1) {
          throw new Error(`--confidence must be a number between 0 and 1, got "${options.confidence}"`);
        }
        node.frontmatter.confidence = confidence;
        changed.push("confidence");
      }

      if (options.status !== undefined) {
        node.frontmatter.status = options.status as typeof node.frontmatter.status;
        changed.push("status");
      }

      if (changed.length === 0) {
        console.log(chalk.yellow("Nothing to update. Pass --content, --tags, --confidence, or --status."));
        return;
      }

      node.frontmatter.modified = nowIso();
      writeNode(paths, node);
      appendChangelog(paths, { action: "update", id, details: { changed } });

      if (options.noIndex !== true && manifest.settings.auto_index) {
        reindexStore(paths);
      }

      console.log(success(`Updated ${chalk.bold(id)}`));
      console.log(kv("changed", changed.join(", ")));
    });
}
